ergo/distributions/point_density.py

```diff
@@ -0,0 +1,260 @@
+            )
+        self.normed_log_densities = init_np.log(self.normed_densities)
+
+    # Distribution
+
+    def pdf(self, x):
+        """
+        If x is out of distribution range, returns 0. Otherwise returns the
+        density at the lowest bin for which the lower bound of the bin
+        is greater than or equal to x.
+
+        :param x: The point in the distribution to get the density at
+        """
+        x = self.scale.normalize_point(x)
+        bin = np.argmin(np.abs(self.normed_xs - x))
+        return np.where(
+            (x < 0) | (x > 1),
+            0,
+            self.scale.denormalize_density(
+                self.scale.denormalize_point(self.normed_xs[bin]),
+                self.normed_densities[bin],
+            ),
+        )
+
+    def logpdf(self, x):
+        return np.log(self.pdf(x))
+
+    def cdf(self, x):
+        normed_x = self.scale.normalize_point(x)
+
+        def in_range_cdf(normed_x):
+            bin = np.argmax(constants.grid > normed_x) - 1
+            c_below_bin = np.where(bin > 0, self.cumulative_normed_ps[bin], 0)
+            c_in_bin = self.normed_densities[bin] * (normed_x - constants.grid[bin])
+            return c_below_bin + c_in_bin
+
+        return np.where(
+            normed_x <= constants.grid[0],
+            0,
+            np.where(normed_x >= constants.grid[-1], 1, in_range_cdf(normed_x)),
+        )
+
+    def ppf(self, q):
+        bin = np.argmin(np.abs(self.cumulative_normed_ps - q))
+        return self.scale.denormalize_point(constants.grid[bin])
+
+    def sample(self):
+        raise NotImplementedError
+
+    # Scaling
+
+    def normalize(self):
+        return PointDensity(
+            self.normed_xs, self.normed_densities, scale=Scale(0, 1), normalized=True
+        )
+
+    def denormalize(self, scale: Scale):
+        return PointDensity(
+            self.normed_xs, self.normed_densities, scale=scale, normalized=True
+        )
+
+    # Structured
+
+    @classmethod
+    def structure(cls, params):
+        class_params, numeric_params = params
+        density_class, scale_classes = class_params
+        density_numeric, scale_numeric = numeric_params
+        return density_class(
+            xs=density_numeric[0],
+            densities=density_numeric[1],
+            scale=scale_classes[0].structure((scale_classes, scale_numeric)),
+            normalized=True,
+            traceable=True,
+            cumulative_normed_ps=density_numeric[2],
+        )
+
+    def destructure(self):
+        scale_classes, scale_numeric = self.scale.destructure()
+        class_params = (self.__class__, scale_classes)
+        self_numeric = self.normed_xs, self.normed_densities, self.cumulative_normed_ps
+        numeric_params = (self_numeric, scale_numeric)
+        return (class_params, numeric_params)
+
+    # Create
+
+    @classmethod
+    def from_pairs(
+        cls, pairs, scale: Scale, normalized=False, allow_non_standard_pairs=True
```

Comment:
1. Are there places where `allow_non_standard_pairs` is `False`?
2. Rename to `allow_nonstandard_pairs`
3. I wonder if we should just let the caller more directly say `interpolate=True` instead of the interpolation check below, and replace the `allow_nonstandard_pairs` argument with the `interpolate` argument.